File: java/BogoSort.java
import java.util.Random;

public class BogoSort {
    public <T extends Comparable<T>> void sort(T[] arr) {
        while (isSorted(arr) == false) {
            shuffle(arr);
        }
    }

    <T extends Comparable<T>> boolean isSorted(T[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (arr[i].compareTo(arr[i - 1]) < 0) {
                return false;
            }
        }
        return true;
    }

    static <T> void swap(T[] arr, int i, int j) {
        T temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    static <T> void shuffle(T[] arr) {
        Random rand = new Random();
        for (int i = 1; i < arr.length; i++) {
            int randomIndex = rand.nextInt(i + 1);
            swap(arr, i, randomIndex);
        }
    }
}

File: src/bogosort.rs
use rand::{Rng, thread_rng};

/// Bogosort returns the number of attempts required to sort the array.
/// Since this is random, the number can vary quite considerably.
/// The f64 array in the examples has taken between 10 and 77,000 attempts in trial runs.
pub fn bogosort<T: PartialOrd>(arr: &mut [T]) -> usize {
    let mut attempts = 0;
    while !is_sorted(arr) {
        shuffle(arr);
        attempts += 1;
    }

    attempts
}

fn is_sorted<T: PartialOrd>(arr: &[T]) -> bool {
    for i in 1..arr.len() {
        if arr[i] < arr[i - 1] {
            return false;
        }
    }
    true
}

fn shuffle<T>(arr: &mut [T]) {
    let mut rng = thread_rng();
    for i in 1..arr.len() {
        let rdm_idx = rng.gen_range(0..=i);
        arr.swap(i, rdm_idx);
    }
}